Add gamepad api hooks

Ledzz opened the pull request.

cc-bbohlender commented:
@Ledzz This looks very good.
Reading your code got me wondering if there might be improvements possible.

The problem is that if a developer wants to `useXRGamepadButton` for two different button, he wouldn't be aware that we are reading all values twice.
I know that this issue would be improved by using a globally defined `Map<id, gamepadValues>` as you mentioned earlier, but I don't want to put this burden on the developer.

Instead, what if the api would be `useXRGamepadReader(inputSource: XRInputSource): { readButtonState(id: string): ButtonState, readAxisState(): { x: number, y: number }, readButton(id: string): number  }`

This way, the developer would call `reader.readButtonState()` in useFrame themselves

Building the `useXRGamepadButton` on top of `useXRGamepadReader` would also be much simpler and only involve reading the exact button.

What are you thinking about such an API. Would it still fit your use case. If not, maybe you can give me some example use cases that don't work or would be harmed from such an API.

cc-bbohlender commented:
@Ledzz Hey, This is a nice implementation too ;-)
And no reason to be sorry. I didn't see a delay :)

My inner perfectionist has 2 comments on the implementation.
1. Performance: I would move the two calls to `const profileComponents = profile?.layouts?.[inputSource.handedness]?.components` outside of the useCallbacks since everything inside the profile.layout[...].components is immutable. I know this sounds like nitpicking, but those functions are called very often so resolving these lookups once and not at each frame could be beneficial.
2. Naming: For me, it’s not yet obvious how the functions on the reader differ. I would suggest `readButton, readButtonValue, readButtonState, readAxis` and rename ComponentState to ButtonState.

I have a third point, which is even more nitpicky, so you can just ignore it if you disagree.
3. Performance: The `readAxisState` function is the only function creating an object and allocating memory. In the three.js style, this could also be solved using `readAxis(id: string, target: Vector2): boolean` returning false if the gamepad or axis doesn’t exist

Again, you are the expert on how you would like to use the API and what makes the most sense, so please let me know if I am wrong :)
Also, since this PR already looks almost done, we can start updating the docs and maybe add an example. It would be awesome if you could also do that so we can get this PR merged ASAP :)

Ledzz commented:
@cc-bbohlender thank you for your politeness, I appreciate it!

Your perfectionism is something I'm lacking sometimes, so your comments are very valuable.

I've updated function names. Please note I've changed it to `readAxes` because if I understand correctly it's a plural form of axes. I'm not a native speaker so feel free to correct me:)

I've also updated the `all-hooks.md`, but I don't know if this explanation is enough.

Also I will add an example when I'll have sometime later, so don't merge this yet, please.

cc-bbohlender commented:
Nice, looks very good, thanks !

Could you add some information to the documentation about what button ids are available for the common controllers (e.g. Quest 1 & 2).

Looking forward to the examples :)
We would love to do a Twitter post about this new feature. Would it be okay to mention you. If you want, you can also do the post yourself, and we'll retweet ;-)

Ledzz commented:
I've updated the docs and the readme with the link to the codesandbox, but it's a work in progress.
Do you have any workflow to use not released yet version of library in codesandbox?

My handle in twitter is `@Ledzz1994`, I don't use it actively but I'd be happy if you mentioned me:)

cc-bbohlender commented:
@Ledzz 
Thanks!
beta/alpha versions are not published automatically yet, so I've quickly made that manually :)
[0.0.33-beta.0](https://www.npmjs.com/package/@coconut-xr/natuerlich/v/0.0.33-beta.0)

As a idea for the example, I could imagine having controller based locomotion (although I am not a fan of this interaction because of motion-sickness, we've received some requestes on its possibility, so a demo on that would be quite nice, but feel free do build anything you prefer :))

cc-bbohlender commented:
Hey @Ledzz , can I assist you with something to get this PR merged? :)

Ledzz commented:
@cc-bbohlender sorry, I was on vacation :) The only thing left is the example, but I personally don't think that it is blocking from merging the PR, that do you think?

cc-bbohlender commented:
@Ledzz no problem, sounds good :)

Ledzz commented:
I think that I've finished the example, could you please check it?
